### ISSUE TITLE
[Bug] features_extractor with net_arch

### ISSUE THREAD

**Issue by Counterfeiter** [DLR-RM/stable-baselines3#310]
### 🐛 Bug

If I use my own features_extractor_class (CNN) with PPO and set the net_arch to some hidden units it will not learn. Also I can't find the source code where features_extractor and policy network combined to train.
If I use only a features_extractor_class and no net_arch it will train. Also If the features_extractor is flatten and I use net_arch.

### To Reproduce

Start PPO training with this options:
policy_kwargs = dict(features_extractor_class=MySimpleFF, activation_fn=nn.ReLU, net_arch=[2, 1])

```python
class MySimpleFF(BaseFeaturesExtractor):

    def __init__(self, observation_space: gym.spaces.Box, features_dim: int = 3, non_conv_channels: int = 2):
        super(MySimpleFF, self).__init__(observation_space, features_dim)
        self.non_conv_channels = observation_space.shape[1] # check if correct for your observations, 0 = batch?
        self.linear = nn.Sequential(
                        nn.Linear(self.non_conv_channels, features_dim), 
                        nn.LeakyReLU(0.1)
        )

    def forward(self, observations: th.Tensor) -> th.Tensor:
        return self.linear(observations)

```

### Expected behavior

Assert If this constellation is not possible

### System Info

Describe the characteristic of your environment:
 * Using conda and git master checkout for stable baselines 3
 * GPU: no
 * Python version: 3.7.x
pytorch                   1.7.1           py3.7_cuda11.0.221_cudnn8.0.5_0    pytorch
torchaudio                0.7.2                      py37    pytorch
torchvision               0.8.2                py37_cu110    pytorch

### Checklist

- [ x] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo (**required**)
- [ x] I have read the [documentation](https://stable-baselines3.readthedocs.io/en/master/) (**required**)
- [ x] I have provided a minimal working example to reproduce the bug (**required**)

**Comment by araffin** [comment]
Hello,
I did a quick check and could not reproduce your issue. I suspect it has to do with the hyperparameters you chose, notably, you should have separated networks for value and actor (cf doc).

Below is two ways of writing the same thing, one with a custom feature extractor, one by passing a `net_arch` argument.
They produce exactly the same results (when using the same seed).

```python
import gym

import torch as th
from stable_baselines3 import PPO
from stable_baselines3.common.torch_layers import BaseFeaturesExtractor


class MySimpleFF(BaseFeaturesExtractor):
    def __init__(self, observation_space: gym.spaces.Box, features_dim: int = 64):
        super(MySimpleFF, self).__init__(observation_space, features_dim)
        obs_dim = observation_space.shape[0]
        self.linear = th.nn.Sequential(
            th.nn.Linear(obs_dim, features_dim),
            th.nn.ReLU(),
        )

    def forward(self, observations: th.Tensor) -> th.Tensor:
        return self.linear(observations)


# 1st way: with custom feature extractor
policy_kwargs = dict(
    features_extractor_class=MySimpleFF,
    net_arch=[dict(pi=[64], vf=[64])],
    activation_fn=th.nn.ReLU,
    ortho_init=False,
)

# 2nd way: with net_arch only
# policy_kwargs = dict(
#     # a first fully connected layer of 64 units
#     # is shared between the actor and critic
#     net_arch=[64, dict(pi=[64], vf=[64])],
#     ortho_init=False,
#     activation_fn=th.nn.ReLU,
# )

model = PPO("MlpPolicy", "CartPole-v1", seed=0, policy_kwargs=policy_kwargs, verbose=1)
model.learn(int(1e5))

```

PS: next time, please add the includes too to reproduce the bug ;)
PPS: we should probably update the doc to recommend more explicitly to use separate networks

**Comment by Counterfeiter** [comment]
Thanks for your quick answer.
Okay, maybe I have a wrong understanding of the concatenation of the layers/networks.

Your example is really perfect to play with it.
Some tests with your example:

I use MySimpleFF with 16 output units for all tests!

    features_extractor_class=MySimpleFF,
    #net_arch=[16],

**Score ~ 150**

    #features_extractor_class=MySimpleFF,
    net_arch=[16],

**Score ~ 150**

    features_extractor_class=MySimpleFF,
    net_arch=[16],

**Score ~ 150**

    #features_extractor_class=MySimpleFF,
    net_arch=[16, 16],

**Score ~ 300**

I thought the last two options are almost equivalent, because of the first two seems to be equivalent?!

My goal is to reduce the final inference time and make the prediction fast with a smaller network. Maybe I miss the fact that learning is worse if the critic or other networks, only used for learning, are to small .

**Comment by Miffyli** [comment]
> I thought the last two options are almost equivalent, because of the first two seems to be equivalent?!

Which `MySimpleFF` you are referring to? If it is the one from first post (not arrafin's), then they are not the same. The `MySimpleFF` has way less units (and also leaky relu instead of tanh activation, which is default for MLP structures), so lower performance is to be expected.

> Maybe I miss the fact that learning is worse if the critic or other networks, only used for learning, are to small .

Yes, if the networks are too small, then you will end up with lower results. Also you won't get speedup from using very small networks as the overhead from calling matrix multiplications and such is much higher than the actual compute (unless you move the network to lower-level language like C).

**Comment by Counterfeiter** [comment]
> Which `MySimpleFF` you are referring to?

Two the one provided by araffin.


Okay, I think I got it. The net_arch was to small and was applied to all networks. So the performance was worse. 

Also I figured out the my goal with a custom 1D CNN (code not provided here) failed, because I added batch normalization. This seems not to work with batch size between 64 and 128. Maybe the batch size should be higher, but for rl its a constrain.
